### PR TITLE
Migrate from Scons to Meson

### PR DESCRIPTION
**Pull request by andy5995**

### CODE REVIEW

**Comment by andy5995** [comment]
At this point, I can't figure out what's causing all the multiple definition errors related to the Surface class

https://github.com/netpanzer/netpanzer/actions/runs/7404731893/job/20146660920#step:6:455

I'll try again later though.

**Comment by andy5995** [comment]
> At this point, I can't figure out what's causing all the multiple definition errors related to the Surface class
> 

Fixed in https://github.com/netpanzer/netpanzer/pull/47

**Comment by andy5995** [comment]
Note the time difference in the CI log. It builds in about 1m30s. With scons it's over 3 minutes.

Though it's not building libphysfs anymore, but linking to an installed version instead.

**Comment by winrid** [comment]
Yeah, scons is much faster now :) I was wondering why it was doing things sequentially.

One of the nice things about scons is that after the work done earlier this year it "just works", I don't normally have to mess around with the build system, even switching from Windows/Linux on a daily basis. I guess that's the bar Meson has to meet :P

**Comment by andy5995** [comment]
> One of the nice things about scons is that after the work done earlier this year it "just works", I don't normally have to mess around with the build system, even switching from Windows/Linux on a daily basis. 

I have doubts that will be the case as more people try to build NetPanzer on more platforms (btw, you can build on bsd and other architectures with [Cirrus CI](https://cirrus-ci.org/). We had to make a few changes recently just to get it to build with clang and honor the CC and CXX environmental variables, something which is normally pretty standard with build systems. Then you had to add a condition so scons wouldn't error if CC/CXX wasn't found in the environment.

But I don't have a Windows system to test on, so I wouldn't really be able to complete this anyway.

As for your comment here regarding the lack of file globbing [> I personally wouldn't want to work with that, seems silly. I don't see a reason to switch from scons to meson if we have to do that.](https://github.com/netpanzer/netpanzer/commit/a961127438156627fd3b1e5cf342dd2165982c04#r136184443)

The reasons for that are at [Why can't I specify target files with a wildcard?](https://mesonbuild.com/FAQ.html#why-cant-i-specify-target-files-with-a-wildcard) though maybe those reasons aren't applicable to the current state of the NetPanzer code base. And like everything else online, everybody has their own opinons on "what's best", lol.

Ok, so I'll quit working on this. Maybe @eli-schwartz will have a little spare time to leave some helpful feedback as well.

**Comment by andy5995** [comment]
For anyone who wants to try this locally, you can run `meson setup _build` and then cd to the _build directory, and enter `meson configure` to see the various build options provided by default the default meson setup.

**Comment by winrid** [comment]
I'm sure lots of people use Meson, it has a ton of Github stars. :) I just don't agree that we still need to list every source file in our build configuration. I am okay to sacrifice a few seconds of initial build time for this, since the project isn't as large as say Postgres.

There are people that clone the repo to their servers and just run "scons" and it works, because they do this to apply custom mods etc, until we support modding. So we have to change their workflow and get them setup with Meson.

Is using Meson blocking you from doing something with project? Then maybe add it for now so you can work, and we can remove it later. I still want to evaluate xmake.

**Comment by winrid** [comment]
Alright, I guess this wins me over: https://www.jetbrains.com/help/clion/meson.html

Let's merge it.

**Comment by andy5995** [comment]
Ok, done. At least a couple things should happen before building with meson is added to the docs.

1. The binary is output to the build root, so it has to be moved to the source root for it to find the data.
    Meson has a "datadir" option so the path can be hard-coded like it is with scons easily enough. But we should discuss how it finds the data when that option isn't given. Either add a `--datadir` option to netpanzer, or have netpanzer look in '../' for the data if it can't find it in the cwd. Probably both should happen. It's pretty customary for games to have a `--datadir` option anyway.

2. Add install targets

I can probably do both of those. I can also add a macos test to the CI.

**Comment by andy5995** [comment]
> Ok, done.

I forgot to do a squash and merge. I'm sorry about that. I'll remember next time...

**Comment by eli-schwartz** [comment]
Another common approach to finding data is to do something like check `$NETPANZER_DATA` as an override directory.

If you do this then meson has `meson.add_devenv()` which allows you to define environment variables that get injected into the shell which is started by `meson devenv -C _build/`.

The meson devenv command automatically sets up the project to build uninstalled, e.g. adding directories with built executables to $PATH, and on Windows (where RPATH doesn't exist) setting up the path so that any `*.dll`s you linked to are able to be loaded. It's also automatically plugged into IDEs such as the vscode plugin: https://github.com/mesonbuild/vscode-meson

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+conf.set_quoted('PACKAGE_VERSION', meson.project_version())
+config_h = configure_file(output : 'config.h', configuration : conf)
+
+deps = [
+  dependency('SDL2'),
+# SDL2 is case sensitive for the ttf library
+  dependency('SDL2_ttf'),
+  dependency('SDL2_mixer'),
+  dependency('physfs')
+]
```
Comment (eli-schwartz) [inline meson.build:39]: Aside: all 4 of these dependencies are available via:
```
meson wrap install ${dep}
```

which means that you can let this build without fuss on Windows where package managers don't exist, and replaces the previous "include a code copy of physfs inside the repository" approach.

**Comment by andy5995** [comment]
Thanks for the tips @eli-schwartz

File: meson.build (hunk 1 above)
Comment (andy5995) [reply meson.build:39]: I'm trying this in https://github.com/netpanzer/netpanzer/pull/133 but the resulting binary always links dynamically to the dependencies. I'm not sure how to handle this so I've opened up a [discussion on the meson repo](https://github.com/mesonbuild/meson/discussions/12754)

**Pull request merged**